File: project_euler_unified/problem_212/solution.py
def solve():
    NCUBOIDS = 50000
    SLEN = 6 * NCUBOIDS

    S = [0] * (SLEN + 1)
    for k in range(1, 56):
        S[k] = (100003 - 200003 * k + 300007 * k * k * k) % 1000000
    for k in range(56, SLEN + 1):
        S[k] = (S[k - 24] + S[k - 55]) % 1000000

    cuboids = []
    for n in range(1, NCUBOIDS + 1):
        x0 = S[6 * n - 5] % 10000
        y0 = S[6 * n - 4] % 10000
        z0 = S[6 * n - 3] % 10000
        dx = 1 + S[6 * n - 2] % 399
        dy = 1 + S[6 * n - 1] % 399
        dz = 1 + S[6 * n] % 399
        cuboids.append((x0, y0, z0, x0 + dx, y0 + dy, z0 + dz))

    zset = set()
    for c in cuboids:
        zset.add(c[2])
        zset.add(c[5])
    zcoords = sorted(zset)

    total_volume = 0

    for zi in range(len(zcoords) - 1):
        z0, z1 = zcoords[zi], zcoords[zi + 1]
        dz = z1 - z0
        active = [c for c in cuboids if c[2] <= z0 and z1 <= c[5]]
        if not active:
            continue

        xset = set()
        for c in active:
            xset.add(c[0])
            xset.add(c[3])
        xcoords = sorted(xset)

        for xi in range(len(xcoords) - 1):
            x0, x1 = xcoords[xi], xcoords[xi + 1]
            dx = x1 - x0
            yintervals = []
            for c in active:
                if c[0] <= x0 and x1 <= c[3]:
                    yintervals.append((c[1], c[4]))
            if not yintervals:
                continue

            yintervals.sort()
            ylen = 0
            cy0, cy1 = yintervals[0]
            for j in range(1, len(yintervals)):
                if yintervals[j][0] >= cy1:
                    ylen += cy1 - cy0
                    cy0, cy1 = yintervals[j]
                else:
                    cy1 = max(cy1, yintervals[j][1])
            ylen += cy1 - cy0

            total_volume += dz * dx * ylen

    print(total_volume)

if __name__ == "__main__":
    solve()

File: project_euler_unified/problem_212/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int NCUBOIDS = 50000;
    const int SLEN = 6 * NCUBOIDS;

    vector<long long> S(SLEN + 1);
    for (int k = 1; k <= 55; k++) {
        long long k3 = (long long)k * k * k;
        S[k] = ((100003LL - 200003LL * k + 300007LL * k3) % 1000000LL + 1000000LL) % 1000000LL;
    }
    for (int k = 56; k <= SLEN; k++)
        S[k] = (S[k - 24] + S[k - 55]) % 1000000LL;

    struct Cuboid { int x1, y1, z1, x2, y2, z2; };
    vector<Cuboid> cuboids(NCUBOIDS);

    for (int n = 1; n <= NCUBOIDS; n++) {
        int i = n - 1;
        cuboids[i] = {
            (int)(S[6*n-5] % 10000), (int)(S[6*n-4] % 10000), (int)(S[6*n-3] % 10000),
            (int)(S[6*n-5] % 10000 + 1 + S[6*n-2] % 399),
            (int)(S[6*n-4] % 10000 + 1 + S[6*n-1] % 399),
            (int)(S[6*n-3] % 10000 + 1 + S[6*n] % 399)
        };
    }

    vector<int> zcoords;
    for (int i = 0; i < NCUBOIDS; i++) {
        zcoords.push_back(cuboids[i].z1);
        zcoords.push_back(cuboids[i].z2);
    }
    sort(zcoords.begin(), zcoords.end());
    zcoords.erase(unique(zcoords.begin(), zcoords.end()), zcoords.end());

    long long totalVolume = 0;

    for (int zi = 0; zi + 1 < (int)zcoords.size(); zi++) {
        int z0 = zcoords[zi], z1 = zcoords[zi + 1];
        int dz = z1 - z0;

        vector<int> active;
        for (int i = 0; i < NCUBOIDS; i++)
            if (cuboids[i].z1 <= z0 && z1 <= cuboids[i].z2)
                active.push_back(i);
        if (active.empty()) continue;

        vector<int> localx;
        for (int i : active) {
            localx.push_back(cuboids[i].x1);
            localx.push_back(cuboids[i].x2);
        }
        sort(localx.begin(), localx.end());
        localx.erase(unique(localx.begin(), localx.end()), localx.end());

        for (int xi = 0; xi + 1 < (int)localx.size(); xi++) {
            int x0 = localx[xi], x1 = localx[xi + 1];
            int dx = x1 - x0;

            vector<pair<int, int>> yintervals;
            for (int i : active)
                if (cuboids[i].x1 <= x0 && x1 <= cuboids[i].x2)
                    yintervals.push_back({cuboids[i].y1, cuboids[i].y2});
            if (yintervals.empty()) continue;

            sort(yintervals.begin(), yintervals.end());
            long long ylen = 0;
            int cy0 = yintervals[0].first, cy1 = yintervals[0].second;
            for (int j = 1; j < (int)yintervals.size(); j++) {
                if (yintervals[j].first >= cy1) {
                    ylen += cy1 - cy0;
                    cy0 = yintervals[j].first;
                    cy1 = yintervals[j].second;
                } else {
                    cy1 = max(cy1, yintervals[j].second);
                }
            }
            ylen += cy1 - cy0;
            totalVolume += (long long)dz * dx * ylen;
        }
    }

    cout << totalVolume << endl;
    return 0;
}
